LabExT/Measurements/InsertionLossSweep.py

```diff
@@ -181,7 +198,57 @@ def algorithm(self, device, data, instruments, parameters):
         self.instr_laser.close()
         self.instr_pm.close()
 
+        # apply reference data
+        if parameters['file path to reference meas.'].value.strip():
+            self.apply_reference_to_data(data=data)
+
         # sanity check if data contains all necessary keys
         self._check_data(data)
 
         return data
+
+    def check_if_reference_valid(self, parameters):
+
+        # load reference file
+        ref_fp = parameters['file path to reference meas.'].value.strip()
+        try:
+            with open(ref_fp, 'r') as fp:
+                ref_raw_data = json.load(fp=fp)
+        except FileNotFoundError:
+            raise FileNotFoundError(f'Reference file "{ref_fp:s}" not found.')
+        except json.decoder.JSONDecodeError:
+            raise RuntimeError(f'Reference file "{ref_fp:s}" could not be JSON decoded.')
+
+        # check reference parameters
+        if ref_raw_data['measurement name'] != self.name:
+            raise ValueError(f'Reference file "{ref_fp:s}" has wrong measurement name, should be {self.name:s}.')
+        check_params = ['wavelength start', 'wavelength stop', 'wavelength step', 'sweep speed', 'laser power']
+        for pname in check_params:
+            if ref_raw_data['measurement settings'][pname]['value'] != parameters[pname].value:
+                raise ValueError(f'Parameters used to record reference in file "{ref_fp:s}" do not match this '
+                                 f"ToDo's settings. Make sure you are using the same settings in the ToDo as "
+                                 f"were used to record the reference file for these parameters: {check_params}.")
+
+        self.ref_data = {
+            'wl': np.array(ref_raw_data['values']['wavelength [nm]']),
+            'tm': np.array(ref_raw_data['values']['transmission [dBm]'])
+        }
+
+    def apply_reference_to_data(self, data):
+
+        assert self.ref_data is not None, 'Reference data is not loaded!'
+
+        # get data back to numpy arrays
+        rec_wl = np.array(data['values']['wavelength [nm]'])
+        rec_tm = np.array(data['values']['transmission [dBm]'])
+
+        # notify user if wavelengths differ > 1pm
+        if any(np.abs(rec_wl - self.ref_data['wl']) > 1e-3):
+            self.logger.warning('Referenced and recorded wavelengths differ by > 1pm!')
+
+        # apply referencing
+        ref_data_diff = rec_tm - self.ref_data['tm']
+        data['values']['referenced transmission [dB]'] = ref_data_diff.tolist()
```

Comment:
Two things that I thought about here:
1. What do you think about the other way around; overwriting data['values']['transmission [dBm]'] with the calibrated transmission (ref_data_diff) and instead storing the "calibration transmission" as an additional value array. Most measurement tools that I know allow you to do a calibration but the calibrated output data does not change it's name. Additionally, all analysis tools and scripts that are based on these value keys would read the calibrated transmission as a default then.
2. What do you think about either making the storage of the calibration / uncalibrated data optional or not storing it all. Is referencing the measurement name / filepath enough? storing this data means approx. 50% of added file storage space. When executing a large number of measurements, this can easily become noticeable.

Reply:
Thanks for the comment. Yes, I am not happy with only conditionally adding the extra output data vector either. I really like your approach 1.

The proposed discarding of the raw measurement data in your option two is risky for the longevity of the data, in my opinion. To make sense, one would have to remember to also save the calibration file that was used to take the measurement with - sometimes a challenge in chaotic research environments.

Based on your feedback, what do you, @boris-vukovic and @delmedico, think about the following changes to this PR:
* when loading a calibration during measurement, store the calibrated transmission in the vector `data['values']['transmission [dBm]']`
  * note that the unit of the referenced transmission is dB, not dBm - so full compatibility with the evaluation script would not be possible anyway
* in case calibration is applied during measurement, make storing the raw power meter values in a separate vector optional (additional boolean parameter)
* reworking this additional functionality into a separate measurement, i.e. not in `InsertionLossSweep`

Reply:
Decisions taken 11.09.2023: @marcoep 
- Don't save different data (calibrated) under the same name of "transmission [dB]". Reason: Confuse people
- Add optional flag "only save calibrated data". Reason: Save disk space
- No new measurement class --> Implement in OG InsertionLossSweep

Reply:
done!

Reply:
@delmedico @boris-vukovic 

Check if you like the new implementation. I added another boolean parameter for the user to selectively discard the raw measurement data.

To note is that all save files now have the three data vectors (`wavelength [nm]`, `transmission [dBm]`, `referenced transmission [dB]`) always present as keys in the `data['values']` dict. If you did not chose to reference the data, the data is an empty list, similarly if you chose to discard the raw data.

Check if you like that solution and I'd be happy for some feedback.